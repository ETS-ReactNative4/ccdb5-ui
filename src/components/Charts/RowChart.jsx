/* eslint complexity: ["error", 5] */

import './RowChart.less'
import * as d3 from 'd3'
import { miniTooltip, row } from 'britecharts'
import { connect } from 'react-redux'
import { hashObject } from '../../utils'
import { max } from 'd3-array'
import PropTypes from 'prop-types'
import React from 'react'
import { toggleTrend } from '../../actions/trends'

export class RowChart extends React.Component {
  _getHeight( numRows ) {
    return numRows === 1 ? 100 : numRows * 60
  }

  _wrapText( text, width ) {
    // ignore test coverage since this is code borrowed from d3 mbostock
    // text wrapping functions
    /* istanbul ignore next */
    // eslint-disable-next-line complexity
    text.each( function() {
      const innerText = d3.select( this )
      if ( innerText.node().children && innerText.node().children.length > 0 ) {
        // assuming its already split up
        return
      }
      const words = innerText.text().split( /\s+/ ).reverse(),
            // ems
            lineHeight = 1.1,
            y = innerText.attr( 'y' ) || 0,
            dy = parseFloat( innerText.attr( 'dy' ) || 0 )

      let word,
          line = [],
          lineNumber = 0,
          tspan = innerText.text( null ).append( 'tspan' ).attr( 'x', 0 )
            .attr( 'y', y ).attr( 'dy', dy + 'em' )

      // eslint-disable-next-line no-cond-assign
      while ( word = words.pop() ) {
        line.push( word )
        tspan.text( line.join( ' ' ) )
        if ( tspan.node().getComputedTextLength() > width ) {
          line.pop()
          tspan.text( line.join( ' ' ) )
          line = [ word ]
          tspan = innerText.append( 'tspan' )
            .attr( 'x', 0 )
            .attr( 'y', y )
            // eslint-disable-next-line no-mixed-operators
            .attr( 'dy', ++lineNumber * lineHeight + dy + 'em' )
            .text( word )
        }
      }
    } )
  }


  componentDidUpdate( prevProps ) {
    const props = this.props
    if( hashObject( prevProps ) !== hashObject( props ) ) {
      this._redrawChart()
    }
  }
  // --------------------------------------------------------------------------
  // Event Handlers
  // eslint-disable-next-line complexity
  _redrawChart() {
    const { colorScheme, data, id, printMode, toggleRow, total } = this.props
    if ( !data || !data.length ) {
      return
    }

    const tooltip = miniTooltip()
    tooltip.valueFormatter( value => value.toLocaleString() + ' complaints' )

    const rows = data
    const ratio = total / max( rows, o => o.value )
    const chartID = '#row-chart-' + id
    d3.selectAll( chartID + ' .row-chart' ).remove()
    const rowContainer = d3.select( chartID )
    const width = printMode ? 750 :
      rowContainer.node().getBoundingClientRect().width
    const height = this._getHeight( rows.length )
    const chart = row()
    const marginLeft = width / 3

    // tweak to make the chart full width at desktop
    // add space at narrow width
    const marginRight = width < 600 ? 20 : -80
    chart.margin( {
      left: marginLeft,
      right: marginRight,
      // left: 200,
      // right: 50,
      top: 20,
      bottom: 10
    } )
      .colorSchema( colorScheme )
      .backgroundColor( '#f7f8f9' )
      // .enableYAxisRight(true) // for trends to show deltas
      .enableLabels( true )
      .labelsTotalCount( total.toLocaleString() )
      .labelsNumberFormat( ',d' )
      .outerPadding( 0.1 )
      .percentageAxisToMaxRatio( ratio )
      .yAxisLineWrapLimit( 2 )
      .yAxisPaddingBetweenChart( 20 )
      .width( width )
      .wrapLabels( true )
      .height( height )
      .on( 'customMouseOver', tooltip.show )
      .on( 'customMouseMove', tooltip.update )
      .on( 'customMouseOut', tooltip.hide )

    rowContainer.datum( rows ).call( chart )
    const tooltipContainer = d3.selectAll( chartID + ' .row-chart .metadata-group' )
    tooltipContainer.datum( [] ).call( tooltip );
    this._wrapText( d3.select( chartID ).selectAll( '.tick text' ), marginLeft )

    rowContainer
      .selectAll( '.y-axis-group .tick' )
      .on( 'click', toggleRow )
  }

  render() {
    return (
      <div className="row-chart-section">
        <h3>{ this.props.title }</h3>
        <div id={ 'row-chart-' + this.props.id }>
        </div>
      </div>
    )
  }
}

export const mapDispatchToProps = dispatch => ( {
  toggleRow: selectedState => {
    dispatch( toggleTrend( selectedState ) )
  }
} )

export const mapStateToProps = ( state, ownProps ) => {
  const { printMode, width } = state.view
  return {
    printMode,
    total: state.aggs.total,
    width
  }
}

RowChart.propTypes = {
  id: PropTypes.string.isRequired,
  colorScheme: PropTypes.oneOfType( [
    PropTypes.array,
    PropTypes.bool
  ] ).isRequired,
  data: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired
}

export default connect( mapStateToProps, mapDispatchToProps )( RowChart )
