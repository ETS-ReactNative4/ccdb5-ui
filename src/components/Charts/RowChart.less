@import (less) "../../css/base.less";

.row-chart-section {
  margin-left: @grid_gutter-width;
  margin-top: @grid_gutter-width/1.5;
}
.row-chart {
  .bg-hover, .visibility, line, path.domain, .x-axis-group {
    display: none;
  }
  text {
    font-size: 16px;
    &.adjust-upwards {
      transform: translateY(-12px);
    }
    &.hidden {
      display:none;
    }
  }
  .view-more-label {
    font-size: 12px;
    text-decoration: underline;
    fill: @pacific;
  }
}
