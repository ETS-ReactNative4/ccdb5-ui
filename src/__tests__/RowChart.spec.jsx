import configureMockStore from 'redux-mock-store'
import { mapStateToProps, RowChart } from '../RowChart'
import { Provider } from 'react-redux'
import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import thunk from 'redux-thunk'


function setupSnapshot() {
  const middlewares = [ thunk ]
  const mockStore = configureMockStore( middlewares )
  const store = mockStore( {
    map: {}
  } )

  return renderer.create(
    <Provider store={ store }>
      <RowChart/>
    </Provider>
  )
}

describe( 'component: RowChart', () => {
  describe( 'initial state', () => {
    it( 'renders without crashing', () => {
      const target = setupSnapshot()
      let tree = target.toJSON()
      expect( tree ).toMatchSnapshot()
    } )
  } )
} )
