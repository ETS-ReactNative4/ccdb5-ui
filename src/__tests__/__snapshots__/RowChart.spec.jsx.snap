// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: RowChart initial state renders without crashing 1`] = `
<div
  className="row-chart-section"
>
  <h3>
    Foo
     by complaint volume (5 highest counts)
  </h3>
  <div
    id="row-chart-foo"
  />
</div>
`;
